Record shape and type in autograd to validate gradients

colesbury opened the pull request:
In the autograd engine, we now validate:

 - The gradient's shape matches the expected shape
 - The gradient's type is compatible with the expected type (i.e. exactly the same or a sparse version of the expected type)

We can also check that gradients are defined when we expect them to be. This check is currently commented out because it triggers in TestJit.test_ge_optimized.

colesbury committed:
Record shape and type in autograd to validate gradients

The check that the gradient is defined is currently disabled because
TestJit.test_ge_optimized will trigger the error.

colesbury committed:
Add tests for new assertions

colesbury committed:
Fix test/cpp/api/misc.cpp

goldsborough reviewed (approved):
Approve for C++ API parts

colesbury committed:
Clear grad_accumulator when Type of Variable.data changes

ezyang commented on torch/csrc/autograd/function.h:
```diff
@@ -123,20 +121,37 @@ struct Function : std::enable_shared_from_this<Function> {
-  uint32_t bump_inputs() noexcept {
-    return num_inputs_++;
+  /// value. Type and shape are the expected properties of the input.
+  uint32_t bump_inputs(const at::Type& type, at::IntList shape) noexcept {
```
A bit of a misnomer now, no?

ezyang commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -288,6 +288,47 @@ static variable_list call_post_hooks(Function& fn, variable_list outputs, variab
   return outputs;
 }
 
+static bool is_compatible_type(const at::Type& expected, const at::Type& actual) {
```
Worth a comment saying that it's OK for a dense tensor to have a sparse gradient, but not vice versa.

ezyang reviewed (approved):
Quality work, though I'm surprised at all the places you needed to add `bump_inputs`, where there wasn't previously a `set_num_inputs`. Intruiging indeed.

colesbury commented on tools/autograd/templates/VariableType.cpp:
```diff
@@ -359,35 +359,35 @@ static void throw_error_out_requires_grad(const char* name) {
 static void rebase_history(Variable& var, std::shared_ptr<Function> grad_fn) {
   if (grad_fn && var.defined()) {
-    grad_fn->set_num_inputs(1);
+    grad_fn->bump_inputs(var.type(), var.sizes());
```
The `grad_fn` is always a new object or null in `rebase_history` and `set_history`. (The `var` may not be new in the case of in-place ops).

colesbury committed:
Rename bump_inputs to add_input_metadata

colesbury committed:
Remove debug code

apaszke commented on torch/csrc/autograd/functions/accumulate_grad.cpp:
```diff
@@ -13,12 +13,13 @@ using at::Tensor;
-              , variable(std::move(variable_)) {}
+    : Function(/*sequence_nr=*/UINT64_MAX)
+    , variable(std::move(variable_)) {
+  add_input_metadata(variable.type(), variable.sizes());
```
Is this correct? `AccumulateGrad` shouldn't return anything, so it doesn't have any inputs

colesbury replied:
The input to AccumulateGrad is the gradient. It has no outputs. (The previous code had `AccumulateGrad(...)  : Function(/*num_inputs=*/1, ...)`

apaszke replied:
Ohhh right. `input_metadata` is for forward outputs (i.e. it's written from the perspective of the backward pass). Never mind the comment

apaszke reviewed (commented):
FWIW, the lack of checks enabled some workarounds that greatly improve the perf for computing a Jacobian (see e.g. #7786) and I'm a bit sad to see it go 😕

colesbury commented:
@apaszke, passing in a larger grad_output only ever worked in trivial cases

apaszke commented:
@colesbury it didn't work with library ops, but the solution was to implement custom Functions in Python that would account for an extra batch dimension in the gradient.

colesbury committed:
Fix type comparison in set_data.

Previously, we were comparing the type of data to the type of this
variable. They were never identical because one was a VariableType and
the other was not.

colesbury commented:
@apaszke I guess we could expose `input_metadata` to Python autograd function if we want to support that sort of trick. i.e. if you want implement those sorts of functions you could call something like `ctx.clear_input_metadata()` to avoid the shape and type checks.

apaszke commented:
@colesbury makes sense. It would be good to have that
